lib/TPP/Transforms/TileConfigHoisting.cpp

```diff
@@ -0,0 +1,101 @@
+//===- TileConfigHoisting.cpp ---------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements tile configuration hoisting on parallel loops.
+//
+//===----------------------------------------------------------------------===//
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_TILECONFIGHOISTINGPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::xsmm;
+
+namespace mlir {
+namespace tpp {
+
+struct TileConfigHoisting : OpRewritePattern<memref::AllocaOp> {
+  using OpRewritePattern<memref::AllocaOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(memref::AllocaOp alloca,
+                                PatternRewriter &rewriter) const override {
+
+    xsmm::TileConfigOp firstTileConfig, secondTileConfig;
+    for (auto *user : alloca->getUsers()) {
+      if (!dyn_cast<xsmm::TileConfigOp>(user)) {
+        return failure();
+      }
+      auto flags =
+          dyn_cast<xsmm::TileConfigDispatchOp>(
+              dyn_cast<xsmm::TileConfigOp>(user).getOperand(0).getDefiningOp())
+              .getFlags();
+      for (auto flagItr : flags) {
+        if (flagItr == xsmm::GemmFlagsAttr::get(
+                           rewriter.getContext(),
+                           mlir::xsmm::GemmFlags::NO_RESET_TILECONFIG)) {
+          firstTileConfig = dyn_cast<xsmm::TileConfigOp>(user);
+
+        } else if (flagItr == xsmm::GemmFlagsAttr::get(
+                                  rewriter.getContext(),
+                                  mlir::xsmm::GemmFlags::NO_SETUP_TILECONFIG)) {
+          secondTileConfig = dyn_cast<xsmm::TileConfigOp>(user);
+        }
+      }
+    }
+
+    scf::ParallelOp parallelOpParent = NULL;
+    auto op = alloca.getOperation();
+    while (true) {
```

Comment:
@alheinecke @rengolin suggested that I flatten this into a two level check instead of a loop, however, as we discussed, the scf fors could be nested deeper than 1 level. should we keep this loop?

Reply:
We don't have any guarantee that we have only 2 scf.for. We have to push up and push down until we reach the scf.parallel. Worst case I can see up to 7 scf.for (for full cache blocking), so there will be any number of 0 to 7 scf.for.